File: year2/week17/solution/week17.java
import java.util.*;
import java.util.function.*;

/*
 * Java 8 style huffman encoding.  Notice that there is not a single
 * null check in the entire program
 */

public class week17 {
    static class HuffmanNode<T> {
        T data;
        int frequency;
        // Use optional to avoid having to null check
        Optional<HuffmanNode> left, right;

        HuffmanNode(T d, int f) {
            data = d;
            frequency = f;
            left = Optional.empty();
            right = Optional.empty();
        }

        HuffmanNode(HuffmanNode l, HuffmanNode r) {
            frequency = l.frequency + r.frequency;
            left = Optional.of(l);
            right = Optional.of(r);
        }

        // Traverse and call f on each leaf node, passing the data
        // and the code
        void traverse(BiConsumer<T, String> f) {
            traverse(f, "");
        }

        void traverse(BiConsumer<T, String> f, String code) {
            if (!left.isPresent() && !right.isPresent())
                f.accept(data, code);
            // if the node is present, traverse it
            left.ifPresent(node -> node.traverse(f, code + "0"));
            right.ifPresent(node -> node.traverse(f, code + "1"));
        }
    }

    // Build a huffman tree given a histogram
    public static <T> HuffmanNode<T> buildTree(HashMap<T, Integer> histogram) {
        PriorityQueue<HuffmanNode<T>> pq =
            new PriorityQueue<>((x, y) -> x.frequency - y.frequency);

        // Insert all values from the histogram into the priority queue
        histogram.forEach((k, v) -> pq.add(new HuffmanNode<T>(k, v)));
        while (pq.size() > 1) {
            HuffmanNode<T> n1 = pq.poll();
            HuffmanNode<T> n2 = pq.poll();

            HuffmanNode<T> parent = new HuffmanNode<>(n1 , n2);
            pq.add(parent);
        }
        return pq.peek();
    }

    public static void main(String args[]) {
        HashMap<String, Integer> histogram = new HashMap<>();
        Scanner stdin = new Scanner(System.in);

        while(stdin.hasNext())
            histogram.merge(stdin.next(), 1, (x, y) -> x + y);
        HuffmanNode<String> root = buildTree(histogram);
        root.traverse((data, code) -> System.out.println(data + " " + code));
    }
}

File: year2/week17/solution/week17.cpp
#include <iostream>
#include <map>
#include <queue>
#include <vector>
#include <functional>

template <typename T>
using histogram = std::map<T, std::size_t>;

template <typename T>
struct huffman_node {
    T data;
    std::size_t frequency;
    std::vector<huffman_node> children;

    huffman_node(const T& d, std::size_t f)
        : data(d), frequency(f) {}

    huffman_node(std::initializer_list<huffman_node> init): frequency(0) {
        children = init;
        for (auto i: init)
            frequency += i.frequency;
    }

    bool operator < (const huffman_node& rhs) const {
        return frequency > rhs.frequency;
    }
};

template <typename T>
histogram<T> build_table(std::istream& is) {
    T token;
    histogram<T> ret;
    while (is >> token)
        ++ret[token];
    return ret;
}

template <typename T>
huffman_node<T> build_tree(std::istream& is) {
    auto table = build_table<T>(is);
    std::priority_queue<huffman_node<T>> pq;

    for (auto i: table)
        pq.push(huffman_node<T>(i.first, i.second));

    while (pq.size() > 1) {
        auto n1 = pq.top(); pq.pop();
        auto n2 = pq.top(); pq.pop();

        huffman_node<T> parent{n1, n2};
        pq.push(parent);
    }
    return pq.top();
}

template <typename T>
using callback_fn = std::function<void(T, std::string)>;

template <typename T>
void traverse(huffman_node<T> root, std::string s, callback_fn<T> cb) {
    if (root.children.empty())
        cb(root.data, s);
    else {
        traverse(root.children[0], s + "0", cb);
        traverse(root.children[1], s + "1", cb);
    }
}

template <typename T>
void traverse(huffman_node<T> root, callback_fn<T> cb) {
    traverse(root, "", cb);
}

int main() {
    auto tree = build_tree<std::string>(std::cin);
    auto cb = [](std::string d, std::string c) {
        std::cout << d << " " << c << std::endl;
    };
    traverse<std::string>(tree, cb);
    return 0;
}
